### ISSUE TITLE
Test packaging in DartPad

### ISSUE THREAD

**Issue by legalcodes** [dart-lang/dart-pad#1120]
Testcode for interactive codelab use cases should be maintainable and readable -- enabling dart's test package would go a long way toward achieving this goal.

**Comment by domesticmouse** [comment]
This requires merging `compile` and `compileDDC` pathways into a newly upgraded `FlutterWebManager`. I'll do some work on this in August, hopefully, but may drag out to Q4.

**Comment by RedBrogdon** [comment]
I've created a test package with some custom code for testing in DartPad. It works right now with dart code, and can be extended for Flutter for web:

https://github.com/redbrogdon/dartpad_test

This package, rather than `test` itself, should be the import goal.

**Comment by legalcodes** [comment]
@galeyang @RedBrogdon My understanding is that we can't use https://github.com/redbrogdon/dartpad_test until https://github.com/dart-lang/dart-pad/issues/901 is resolved, is that correct?

**Comment by domesticmouse** [comment]
I'm happy enough to directly add code into our project that we control as a stop gap before we figure out how to do #901 properly. 

@RedBrogdon do you want to give me an explanation and/or a PR that explains what I need to add to `dart-services` to enable @legalcodes to use your `dartpad_test` package? I think my preference would be if you actually published `dartpad_test` to pub so I can just add it as always present dependency in the `pubspec.yaml`.

**Comment by RedBrogdon** [comment]
The code has been written for some time, waiting on our ability to pull in packages:

https://github.com/redbrogdon/dartpad_test

I haven't yet published it, but certainly can.

**Comment by domesticmouse** [comment]
Let me know when it is published and I’ll integrate into the pubspec.

**Comment by matthew-carroll** [comment]
@RedBrogdon would you mind summarizing the state of writing Dart tests in DartPad? I couldn't quite follow the thread.

I'm providing training services to companies and one of the sections involves Dart tests and widget tests. I'd love to be able to run both of these through DartPad. It sounds like a form of Dart tests might be available here?

**Comment by RedBrogdon** [comment]
Sure thing; I worked on this a good bit last summer. 

**tl;dr**: The code for the testing package is at https://github.com/redbrogdon/dartpad_test. Some examples that use it are at https://github.com/RedBrogdon/dartpad_exercises. The package is currently unpublished, waiting on package support in dart-services. You can, however, lift out parts of the code and include them directly in the test methods for your exercises, where they'll work (though you'd need to avoid relying on package:matcher).

**History**:
The first thing I tried was just running the Dart test runner in a browser environment and trying to hook into it. This turned out to be a dead end, since the test runner was never designed for that kind of workflow. There's nowhere to conveniently cut-in and manipulate/report on the process.

After that, I talked a good bit with @grouma about which way to go, and decided to create my own "test" methods that followed the same API used for typical Dart/Flutter tests and used the same Matchers that everyone's come to know and love. There was a hitch with AsyncMatcher, which relies on the test runner to wait for all the async events to complete before closing down the test, so I was forced to write my own.

The end result is a pair of packages that manage to replicate the test API in a "good enough" way for DartPad, while still staying small. I built the dartpad_exercises repo as a way of checking that the tests would work both in the browser and on the command line for CI. Neither are currently published, as dart-services still lacks support for packages beyond those that ship with the Flutter SDK. We're dead set on adding that, but haven't had bandwidth to do so.

**Comment by RedBrogdon** [comment]
I should add, @matthew-carroll, that I'm happy to set up a half-hour GVC or something if you'd like to talk about this. I'd be interested to hear about your project and where DartPad could fit in.

**Comment by domesticmouse** [comment]
@RedBrogdon you were going to publish `dartpad_test` so we can add it to a list of whitelisted packages before we get to general package support. Or did I misunderstand a thread somewhere?

**Comment by RedBrogdon** [comment]
No, you're correct. I've just been pushing that task to the backburner for several weeks now in favor of other things.

I'll find time this week to dust off the code, get travis set up for that repo, and get it published.

**Comment by parlough** [comment]
As part of the [new, simplified version](https://github.com/dart-lang/dart-pad/issues/2808) of the DartPad frontend, the exercise functionality this was related to was removed.

If anyone would still like to see similar functionality, please check out other related issues or open a new one with your use case. Thanks so much!
